File: ruby/lib/runapi/qwen_2/resources/remix_image.rb
# frozen_string_literal: true

module RunApi
  module Qwen2
    module Resources
      # Qwen2 remix-image resource. Creates prompt-guided variations from a source image.
      class RemixImage
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/qwen_2/remix_image"

        RESPONSE_CLASS = Types::RemixImageResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedRemixImageResponse

        def initialize(http)
          @http = http
        end

        # Remix an image and wait until complete.
        #
        # @param params [Hash] remix-image parameters
        # @return [RunApi::Qwen2::Types::CompletedRemixImageResponse] completed remix-image result
        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        # Create a remix-image task.
        #
        # @param params [Hash] remix-image parameters
        # @return [RunApi::Qwen2::Types::RemixImageResponse] task creation result with id
        def create(**params)
          params = compact_params(params)
          validate_contract!(CONTRACT["remix-image"], params)
          request(:post, ENDPOINT, body: params)
        end

        # Get remix-image status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Qwen2::Types::RemixImageResponse] current remix-image status
        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end
      end
    end
  end
end

File: python/src/runapi/qwen_2/resources/remix_image.py
"""Qwen 2 remix-image resource."""

from __future__ import annotations

from typing import Any

from runapi.core import Resource

from ..contract_gen import CONTRACT
from ..types import (
    CompletedRemixImageResponse,
    RemixImageResponse,
)


class RemixImage(Resource):
    """Create prompt-guided variations from a source image."""

    ENDPOINT = "/api/v1/qwen_2/remix_image"

    RESPONSE_CLASS = RemixImageResponse
    COMPLETED_RESPONSE_CLASS = CompletedRemixImageResponse

    def run(self, **params: Any) -> Any:
        """Create a remix-image task and poll until it completes.

        Args:
            **params: Remix-image parameters (model, prompt, source_image_url, ...).

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create a remix-image task and return immediately with an ``id``.

        Args:
            **params: Remix-image parameters (model, prompt, source_image_url, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["remix-image"], compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of a remix-image task.

        Args:
            id: The task id.

        Returns:
            The current status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")
